pkg/network/tracer/process_cache.go

```diff
@@ -0,0 +1,300 @@
+		p := value.(*process)
+		pl, _ := pc.cacheByPid[p.Pid]
+		if pl = pl.remove(p); len(pl) == 0 {
+			delete(pc.cacheByPid, p.Pid)
+			return
+		}
+
+		pc.cacheByPid[p.Pid] = pl
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	go func() {
+		for {
+			select {
+			case <-pc.stopped:
+				return
+			case p := <-pc.in:
+				pc.add(p)
+			}
+		}
+	}()
+
+	return pc, nil
+}
+
+func (pc *processCache) handleProcessEvent(entry *smodel.ProcessCacheEntry) {
+
+	select {
+	case <-pc.stopped:
+		return
+	default:
+	}
+
+	p := pc.processEvent(entry)
+	if p == nil {
+		pc.stats.eventsSkipped.Add(1)
+		return
+	}
+
+	select {
+	case pc.in <- p:
+	default:
+		// dropped
+		pc.stats.eventsDropped.Add(1)
+	}
+}
+
+func (pc *processCache) processEvent(entry *smodel.ProcessCacheEntry) *process {
+	var envs map[string]string
+	if entry.EnvsEntry != nil {
+		for _, v := range entry.EnvsEntry.Values {
+			kv := strings.SplitN(v, "=", 2)
+			k := kv[0]
+
+			if len(pc.filteredEnvs) > 0 {
+				if _, found := pc.filteredEnvs[k]; !found {
+					continue
+				}
+			}
+
+			v = ""
+			if len(kv) > 1 {
+				v = kv[1]
+			}
+
+			if envs == nil {
+				envs = make(map[string]string)
+			}
+			envs[k] = v
+			if len(pc.filteredEnvs) > 0 && len(pc.filteredEnvs) == len(envs) {
+				break
+			}
+		}
+	}
+
+	if len(envs) == 0 && len(pc.filteredEnvs) > 0 && entry.ContainerID == "" {
+		return nil
+	}
+
+	return &process{
+		Pid:         entry.Pid,
+		Envs:        envs,
+		ContainerID: entry.ContainerID,
+		StartTime:   entry.ExecTime.UnixNano(),
+	}
+}
+
+func (pc *processCache) Stop() {
+	if pc == nil {
+		return
+	}
+
+	pc.stop.Do(func() { close(pc.stopped) })
+}
+
+func (pc *processCache) add(p *process) {
+	if pc == nil {
+		return
+	}
+
+	pc.Lock()
+	defer pc.Unlock()
+
+	evicted := pc.cache.Add(processCacheKey{pid: p.Pid, startTime: p.StartTime}, p)
+	pl, _ := pc.cacheByPid[p.Pid]
+	pc.cacheByPid[p.Pid] = pl.update(p)
+
+	if evicted {
+		pc.stats.cacheEvicts.Add(1)
+	}
+}
+
+func (pc *processCache) GetStats() map[string]interface{} {
+	if pc == nil {
+		return map[string]interface{}{}
+	}
+
+	pc.Lock()
+	defer pc.Unlock()
+
+	pc.stats.cacheLength.Store(uint64(pc.cache.Len()))
+	return atomicstats.Report(&pc.stats)
+}
+
+func (pc *processCache) Get(pid uint32, ts int64) (*process, bool) {
+	if pc == nil {
+		return nil, false
+	}
+
+	pc.Lock()
+	defer pc.Unlock()
+
+	pl, _ := pc.cacheByPid[pid]
+	if closest := pl.closest(ts); closest != nil {
+		pc.cache.Get(processCacheKey{pid: closest.Pid, startTime: closest.StartTime})
+		return closest, true
+	}
+
+	return nil, false
+}
+
+func (pc *processCache) Dump() (interface{}, error) {
+	res := map[uint32]interface{}{}
+	if pc == nil {
+		return res, nil
+	}
+
+	pc.Lock()
+	defer pc.Unlock()
+
+	for pid, pl := range pc.cacheByPid {
+		res[pid] = pl
+	}
+
+	return res, nil
+}
+
+func (pl processList) update(p *process) processList {
+	for i := range pl {
+		if pl[i].Pid == p.Pid && pl[i].StartTime == p.StartTime {
```

Comment:
The `processList` is already filtered by PID right?

Reply:
Right, fixed